File: backend/types/restaurant.py
from dataclasses import asdict, dataclass
from datetime import datetime
from enum import Enum
from typing import Any, Dict, List, Optional, Union

#!/usr/bin/env python3
"""Restaurant Type Definitions for JewGo Backend.
============================================

Unified type definitions for restaurant data structures to ensure
consistency between frontend and backend.

Author: JewGo Development Team
Version: 1.0
"""


class KosherCategory(str, Enum):
    """Kosher category enumeration."""

    MEAT = "meat"
    DAIRY = "dairy"
    PAREVE = "pareve"


class RestaurantStatus(str, Enum):
    """Restaurant status enumeration."""

    OPEN = "open"
    CLOSED = "closed"
    UNKNOWN = "unknown"


class SpecialType(str, Enum):
    """Special type enumeration."""

    DISCOUNT = "discount"
    PROMOTION = "promotion"
    EVENT = "event"


@dataclass
class RestaurantSpecial:
    """Restaurant special offer data structure."""

    id: int
    restaurant_id: int
    title: str
    description: str | None = None
    discount_percent: float | None = None
    discount_amount: float | None = None
    start_date: str | None = None
    end_date: str | None = None
    is_paid: bool = False
    payment_status: str = "pending"
    special_type: SpecialType = SpecialType.DISCOUNT
    priority: int = 1
    is_active: bool = True
    created_date: str | None = None
    updated_date: str | None = None


@dataclass
class MenuPricing:
    """Menu pricing data structure."""

    min: float
    max: float
    avg: float


@dataclass
class HoursData:
    """Hours of operation data structure."""

    day: str
    open_time: str | None = None
    close_time: str | None = None
    is_closed: bool = False
    is_24_hours: bool = False


@dataclass
class Restaurant:
    """Unified restaurant data structure."""

    # Core identification
    id: int
    name: str

    # Location information
    address: str
    city: str
    state: str
    zip_code: str
    latitude: float | None = None
    longitude: float | None = None

    # Contact information
    phone_number: str
    website: str | None = None
    google_listing_url: str | None = None

    # Kosher certification
    kosher_category: KosherCategory
    certifying_agency: str = "ORB"
    is_cholov_yisroel: bool | None = None
    is_pas_yisroel: bool | None = None

    # Business details
    listing_type: str
    short_description: str | None = None
    price_range: str | None = None
    min_avg_meal_cost: float | None = None
    max_avg_meal_cost: float | None = None

    # Hours and status
    hours_of_operation: str | None = None
    hours_json: str | None = None
    hours_last_updated: str | None = None
    timezone: str | None = None
    current_time_local: str | None = None
    hours_parsed: bool = False

    # Status information
    status: RestaurantStatus = RestaurantStatus.UNKNOWN
    is_open: bool | None = None
    status_reason: str | None = None
    next_open_time: str | None = None

    # Media
    image_url: str | None = None

    # Specials and offers
    specials: list[RestaurantSpecial] = None

    # Ratings and reviews
    rating: float | None = None
    star_rating: float | None = None
    quality_rating: float | None = None
    review_count: int | None = None
    google_rating: float | None = None
    google_review_count: int | None = None
    google_reviews: str | None = None

    # Timestamps
    created_at: str | None = None
    updated_at: str | None = None

    # Additional notes
    notes: str | None = None

    def __post_init__(self):
        """Post-initialization processing."""
        if self.specials is None:
            self.specials = []

        # Ensure kosher_category is an enum
        if isinstance(self.kosher_category, str):
            self.kosher_category = KosherCategory(self.kosher_category)

        # Ensure status is an enum
        if isinstance(self.status, str):
            self.status = RestaurantStatus(self.status)

    def to_dict(self) -> dict[str, Any]:
        """Convert to dictionary format for API responses."""
        data = asdict(self)

        # Convert enums to strings for JSON serialization
        data["kosher_category"] = self.kosher_category.value
        data["status"] = self.status.value

        # Convert specials to dictionaries
        if self.specials:
            data["specials"] = [
                special.to_dict() if hasattr(special, "to_dict") else asdict(special)
                for special in self.specials
            ]

        return data

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "Restaurant":
        """Create Restaurant instance from dictionary."""
        # Handle specials conversion
        if data.get("specials"):
            data["specials"] = [
                RestaurantSpecial(**special) if isinstance(special, dict) else special
                for special in data["specials"]
            ]

        return cls(**data)

    def get_rating(self) -> float | None:
        """Get the best available rating."""
        return self.rating or self.star_rating or self.google_rating

    def get_price_range(self) -> str:
        """Get formatted price range."""
        if self.price_range:
            return self.price_range

        if self.min_avg_meal_cost and self.max_avg_meal_cost:
            return f"${self.min_avg_meal_cost} - ${self.max_avg_meal_cost}"

        return "Price not available"

    def is_kosher_certified(self) -> bool:
        """Check if restaurant has kosher certification."""
        return bool(self.certifying_agency and self.certifying_agency != "None")

    def get_kosher_details(self) -> dict[str, Any]:
        """Get kosher certification details."""
        details = {
            "category": self.kosher_category.value,
            "agency": self.certifying_agency,
            "is_certified": self.is_kosher_certified(),
        }

        if (
            self.kosher_category == KosherCategory.DAIRY
            and self.is_cholov_yisroel is not None
        ):
            details["is_cholov_yisroel"] = self.is_cholov_yisroel

        if (
            self.kosher_category in [KosherCategory.MEAT, KosherCategory.PAREVE]
            and self.is_pas_yisroel is not None
        ):
            details["is_pas_yisroel"] = self.is_pas_yisroel

        return details


# Type aliases for convenience
RestaurantDict = dict[str, Any]
RestaurantList = list[Restaurant]
RestaurantDictList = list[RestaurantDict]

# Validation schemas
RESTAURANT_REQUIRED_FIELDS = [
    "name",
    "address",
    "city",
    "state",
    "zip_code",
    "phone_number",
    "kosher_category",
    "listing_type",
]

KOSHER_CATEGORIES = [cat.value for cat in KosherCategory]
RESTAURANT_STATUSES = [status.value for status in RestaurantStatus]

File: frontend/lib/types/restaurant.ts
// Kosher category enumeration
export type KosherCategory = 'Meat' | 'Dairy' | 'Pareve' | 'Fish' | 'Unknown';

// Restaurant status enumeration
export type RestaurantStatus = 'open' | 'closed' | 'unknown';

// Special type enumeration
export type SpecialType = 'discount' | 'promotion' | 'event';

// Hours structure interfaces
export interface DayHours {
  open: string;
  close: string;
  closed?: boolean;
}

export interface WeeklyHours {
  monday?: DayHours;
  tuesday?: DayHours;
  wednesday?: DayHours;
  thursday?: DayHours;
  friday?: DayHours;
  saturday?: DayHours;
  sunday?: DayHours;
}

export interface HoursData {
  hours?: WeeklyHours;
  hours_json?: string;
  hours_of_operation?: string;
  hours_last_updated?: string;
  timezone?: string;
  current_time_local?: string;
  hours_parsed?: boolean;
}

// Category structure interface
export interface RestaurantCategory {
  id?: number;
  name: string;
  description?: string;
  icon?: string;
  color?: string;
}

export interface RestaurantSpecial {
  id: number;
  restaurant_id: number;
  title: string;
  description?: string;
  discount_percent?: number;
  discount_amount?: number;
  start_date?: string;
  end_date?: string;
  is_paid: boolean;
  payment_status: string;
  special_type: 'discount' | 'promotion' | 'event';
  priority: number;
  is_active: boolean;
  created_date: string;
  updated_date: string;
  // Additional properties for UI display
  discount?: string;
  price?: number;
  originalPrice?: number;
  validUntil?: string;
  terms?: string;
}

export interface MenuPricing {
  [section: string]: {
    min: number;
    max: number;
    avg: number;
  };
}

export interface RestaurantCardProps {
  restaurant: Restaurant;
  className?: string;
}

export interface RestaurantGridProps {
  restaurants: Restaurant[];
  currentPage?: number;
  totalPages?: number;
  onPageChange?: (page: number) => void;
  totalRestaurants?: number;
}

export interface SearchBarProps {
  onSearch: (query: string) => void;
}

export interface CategoryNavProps {
  selectedCategory: string;
  onCategoryChange: (category: string) => void;
}

export interface Restaurant {
  // Hours and category with proper types
  hours: HoursData;
  category: RestaurantCategory;
  
  // Core identification
  id: string;
  name: string;
  
  // Location information
  address: string;
  city: string;
  state: string;
  zip_code: string;
  latitude?: number;
  longitude?: number;
  
  // Contact information
  phone_number: string;
  website?: string;
  google_listing_url?: string;
  
  // Kosher certification
  kosher_category: KosherCategory;
  certifying_agency: string;
  is_cholov_yisroel?: boolean;
  is_pas_yisroel?: boolean;
  cholov_stam?: boolean;
  
  // Business details
  listing_type: string;
  short_description?: string;
  price_range?: string;
  min_avg_meal_cost?: number;
  max_avg_meal_cost?: number;
  
  // Hours and status
  hours_of_operation?: string;
  hours_json?: string;
  hours_last_updated?: string;
  timezone?: string;
  current_time_local?: string;
  hours_parsed?: boolean;
  
  // Status information
  status: RestaurantStatus;
  is_open?: boolean;
  status_reason?: string;
  next_open_time?: string;
  
  // Media
  image_url?: string;
  additional_images?: string[];
  
  // Specials and offers
  specials?: RestaurantSpecial[];
  
  // Ratings and reviews
  rating?: number;
  star_rating?: number;
  quality_rating?: number;
  review_count?: number;
  google_rating?: number;
  google_review_count?: number;
  google_reviews?: string;
  
  // Enhanced business data (Phase 3)
  business_types?: string;
  review_snippets?: string;
  
  // Timestamps
  created_at?: string;
  updated_at?: string;
  
  // Additional notes
  notes?: string;
  
  // Legacy fields for backward compatibility
  certificate_link?: string;
  hours_open?: string;
  avg_price?: string;
  menu_pricing?: MenuPricing;
  openStatus?: RestaurantStatus;
  
  // Additional fields that may be present in data
  dietary_restrictions?: string[];
  
  // Computed fields (not from database)
  distance?: string;
}

// Utility function to compute restaurant status
export const computeRestaurantStatus = (restaurant: Restaurant): 'open' | 'closed' | 'unknown' => {
  // Priority: is_open field first
  if (restaurant.is_open !== undefined) {
    return restaurant.is_open ? 'open' : 'closed';
  }
  
  // Fallback to hours_open field
  if (restaurant.hours_open) {
    return restaurant.hours_open.toLowerCase().includes('open') ? 'open' : 'closed';
  }
  
  // Fallback to status field
  if (restaurant.status) {
    const status = restaurant.status.toLowerCase();
    if (status.includes('open')) {return 'open';}
    if (status.includes('closed')) {return 'closed';}
  }
  
  return 'unknown';
};
